### PR TITLE
ci: add typo detection tool

### PR DESCRIPTION
**Pull request by bodhi-crypo**
# Description

- add automatic typo detection in CI

## Additional Notes

- Blocks all CHANGELOG typos checks, although there are indeed some typos in it, it does not affect the readability of the main code.
- If we need to check CHANGELOG, we need to add further extend-words eg.`ede`
```
error: `ede` should be `edge`
  --> ./domainobjs/CHANGELOG.md:56:64
   |
56 | * **deps-dev:** bump @types/node from 20.11.2 to 20.11.5 ([7698ede](https://github.com/privacy-scaling-explorations/maci/commit/7698edef2b95bc017e54c1fc0d796ccc9da3de85))
   |                                                                ^^^
   |
error: `polution` should be `pollution`
  --> ./crypto/CHANGELOG.md:27:46
   |
27 | * **crypto:** get rid of potential prototype polution ([9b8a97c](https://github.com/privacy-scaling-explorations/maci/commit/9b8a97cc9661e0ee6f535cdcab19b4aa0fa98136))
   |                                              ^^^^^^^^
   |
error: `ede` should be `edge`
  --> ./crypto/CHANGELOG.md:65:64
   |
65 | * **deps-dev:** bump @types/node from 20.11.2 to 20.11.5 ([7698ede](https://github.com/privacy-scaling-explorations/maci/commit/7698edef2b95bc017e54c1fc0d796ccc9da3de85))
```

## Related issue(s)

Original issue: https://github.com/privacy-scaling-explorations/halo2curves/issues/142

## Confirmation

- [x] I have read and understand MACI's [contributor guidelines](https://maci.pse.dev/docs/contributing) and [code of conduct](https://maci.pse.dev/docs/contributing/code-of-conduct).
- [x] I have read and understand MACI's [GitHub processes](https://github.com/privacy-scaling-explorations/maci/discussions/847).
- [x] I have read and understand MACI's [testing guide](https://maci.pse.dev/docs/testing).

### CODE REVIEW

**Comment by ctrlc03** [comment]
> # Description
> * add automatic typo detection in CI
> 
> ## Additional Notes
> * Blocks all CHANGELOG typos checks, although there are indeed some typos in it, it does not affect the readability of the main code.
> * If we need to check CHANGELOG, we need to add further extend-words eg.`ede`
> 
> ```
> error: `ede` should be `edge`
>   --> ./domainobjs/CHANGELOG.md:56:64
>    |
> 56 | * **deps-dev:** bump @types/node from 20.11.2 to 20.11.5 ([7698ede](https://github.com/privacy-scaling-explorations/maci/commit/7698edef2b95bc017e54c1fc0d796ccc9da3de85))
>    |                                                                ^^^
>    |
> error: `polution` should be `pollution`
>   --> ./crypto/CHANGELOG.md:27:46
>    |
> 27 | * **crypto:** get rid of potential prototype polution ([9b8a97c](https://github.com/privacy-scaling-explorations/maci/commit/9b8a97cc9661e0ee6f535cdcab19b4aa0fa98136))
>    |                                              ^^^^^^^^
>    |
> error: `ede` should be `edge`
>   --> ./crypto/CHANGELOG.md:65:64
>    |
> 65 | * **deps-dev:** bump @types/node from 20.11.2 to 20.11.5 ([7698ede](https://github.com/privacy-scaling-explorations/maci/commit/7698edef2b95bc017e54c1fc0d796ccc9da3de85))
> ```
> 
> ## Related issue(s)
> Original issue: [privacy-scaling-explorations/halo2curves#142](https://github.com/privacy-scaling-explorations/halo2curves/issues/142)
> 
> ## Confirmation
> * [x]  I have read and understand MACI's [contributor guidelines](https://maci.pse.dev/docs/contributing) and [code of conduct](https://maci.pse.dev/docs/contributing/code-of-conduct).
> * [x]  I have read and understand MACI's [GitHub processes](https://github.com/privacy-scaling-explorations/maci/discussions/847).
> * [x]  I have read and understand MACI's [testing guide](https://maci.pse.dev/docs/testing).

Hi @bodhi-crypo thanks for this PR. could you please run prettier on this branch and rebase to our latest dev changes please?

**Comment by bodhi-crypo** [comment]
> rebase

@ctrlc03 rebased

**Comment by ctrlc03** [comment]
> > rebase
> 
> @ctrlc03 rebased

Could you run prettier as well please? `pnpm prettier:write` 🙏

**Comment by bodhi-crypo** [comment]
> > > rebase
> > 
> > 
> > @ctrlc03 rebased
> 
> Could you run prettier as well please? `pnpm prettier:write` 🙏

I know why the preitter failed. The newadded typos.yaml has a format problem, actually need run `pnpm prettier:fix`

File: .github/workflows/.config/typos.toml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[default.extend-words]
+ba = "ba"
+groth = "groth"
+
+[files]
+extend-exclude = ["/**/CHANGELOG.md"]
```
Comment (0xmad) [inline .github/workflows/.config/typos.toml:6]: Please, add new line at the end of the file

File: .github/workflows/typos.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+jobs:
+  typos:
+    name: Spell Check with Typos
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - name: Use typos with config file
+        uses: crate-ci/typos@master
+        with:
+          config: .config/typos.toml
```
Comment (0xmad) [inline .github/workflows/typos.yml:14]: Let's use `config` folder instead of `.config`

File: .github/workflows/typos.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+name: Typo Check
+
+on: [pull_request]
```
Comment (0xmad) [inline .github/workflows/typos.yml:3]:
```suggestion
on:
  push:
    branches: [dev]
  pull_request:

concurrency:
  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
  cancel-in-progress: true

```

File: .github/workflows/typos.yml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+jobs:
+  typos:
+    name: Spell Check with Typos
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - name: Use typos with config file
+        uses: crate-ci/typos@master
```
Comment (0xmad) [inline .github/workflows/typos.yml:12]: It should be a fixed tag version, not latest code from master

**Comment by bodhi-crypo** [comment]
> @bodhi-crypo thanks, just left some comments

fixed && squashed @ctrlc03

**Review by 0xmad — APPROVED** [review]
@bodhi-crypo thanks!

**Pull request merged**